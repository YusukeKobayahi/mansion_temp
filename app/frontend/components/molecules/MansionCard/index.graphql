fragment MansionCardData on Mansion {
  uniqueCode
  name
  age
  constructedIn
  unitAmount
  pageView
  prefecture {
    name
  }
  city {
    name
  }
  town {
    name
  }
  street {
    name
  }
  blockNumber
  display(md: $md, prb: $prb, prt: $prt, eab: $eab, eat: $eat) {
    minPrice
    maxPrice
    minExclusiveArea
    maxExclusiveArea
    layoutName
  }
  access {
    id
    stationId
    stationLineName
    stationName
    stationWalkingMinutes
    lineId
  }
  ...MsLibraryMansionsUnitSize
}
