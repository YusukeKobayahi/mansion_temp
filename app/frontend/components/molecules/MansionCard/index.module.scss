@import "~/styles/variables";
@import "~/styles/utils";

.container {
  border: 1px solid #e5e5e5;
  border-radius: 4px;
  height: 100%;

  .pcOnly {
    height: 100%;
  }

  .spOnly {
    display: none;
  }

  @include screenMd {
    padding: 0;

    .pcOnly {
      display: none;
    }

    .spOnly {
      display: block;
    }
  }
}

.name {
  min-height: 40px;
  background-color: #f8f6f0;
  padding: 10px 30px 10px 10px;
  color: #b2a173;
  font-size: 18px;
  position: relative;
  text-decoration: underline;
  display: flex;
  align-items: center;
  line-height: 1.5;

  @include screenMd {
    line-height: 1.5;
    background-color: #ffffff;
  }

  &::after {
    content: "";
    position: absolute;
    top: 50%;
    right: 15px;
    width: 10px;
    height: 10px;
    border-top: 1px solid #b2a173;
    border-right: 1px solid #b2a173;
    transform: translateY(-50%) rotate(45deg);
  }
}

.columns {
  width: 100%;
  padding: 30px 15px 0;
  margin-bottom: 30px;

  @include screenMd {
    display: block;
  }
}

.images {
  width: 200px;
  height: 150px;
  margin: 0 auto;
  background-color: #cccccc;
}

.block {
  display: flex;
  flex-direction: column;
  padding-bottom: 20px;
  height: 100%;
}

.table {
  .item {
    border-top: 1px solid #e5e5e5;
    display: flex;
    align-items: center;
    padding: 15px 0;

    p {
      font-size: 14px;
      color: #373a3c;
      &.label {
        min-width: 110px;
        margin-right: 10px;
      }
      &.content {
        span {
          display: block;

          &.highlight {
            font-weight: bold;
          }
        }
      }
    }

    &:nth-last-child(1) {
      border-bottom: 1px solid #e5e5e5;
    }
  }
}

.description {
  margin-top: 15px;

  p {
    font-size: 14px;
    color: #373a3c;
    line-height: 1.5;

    &.descriptionTitle {
      font-weight: bold;
    }

    &.descriptionContent {
      padding-left: 15px;
    }
  }
}

button.link {
  width: 285px;
  display: block;
  margin: auto auto 0;
  line-height: 50px;
  border: 1px solid #bca565;
  color: #bca565;
  text-align: center;
  font-size: 16px;
  padding: 0;
  background-color: unset;

  @include screenMd {
    width: unset;
  }
}

.card {
  @include screenMd {
    width: 100%;
    padding: 16px 35px 16px 20px;
    display: block;
    position: relative;
    background-color: unset;
    text-align: left;

    &::after {
      content: "";
      position: absolute;
      top: 50%;
      right: 20px;
      width: 10px;
      height: 10px;
      border-top: 1px solid #b2a173;
      border-right: 1px solid #b2a173;
      transform: translateY(-50%) rotate(45deg);
    }
    h3 {
      color: #b2a173;
      font-size: 18px;
      position: relative;
      text-decoration: underline;
    }

    ul {
      margin-top: 15px;
      li {
        font-size: 15px;
        line-height: 1.5;
        color: #646464;
      }
    }

    p {
      margin-top: 15px;
      font-size: 14px;
      color: #9d9d9d;
    }
  }
}

.page_view {
  padding: 0px 15px;
  margin-bottom: 30px;
  font-size: 14px;
  color: #373a3c;
}
