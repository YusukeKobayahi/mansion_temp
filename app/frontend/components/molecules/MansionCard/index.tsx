import Link from "next/link";
import classnames from "classnames";
import { event } from "~/lib/gtag";
import { MansionCardDataFragment } from "~/graphql/generated";
import { yenFormat, buildLibraryPagePath } from "~/lib/utils";
import uniq from "lodash/uniq";
import compact from "lodash/compact";
import sortBy from "lodash/sortBy";
import { format } from "date-fns";
import styles from "./index.module.scss";
import tableStyles from "~/components/pages/cities/Info/index.module.scss";
import _Link from "~/components/commons/Link";
import { GoogleAnalyticsEvent } from "~/lib/types";

/**
 * 将来的にimage,reviewなどが拡張される予定
 */
type MansionCardProps = MansionCardDataFragment & {
  subjectNames?: string[];
  writing?: boolean;
  eventParams?: GoogleAnalyticsEvent;
};

const MansionCard: React.FC<MansionCardProps> = ({
  uniqueCode,
  name,
  age,
  constructedIn,
  unitAmount,
  pageView,
  prefecture,
  city,
  town,
  street,
  blockNumber,
  display,
  access,
  salesHistorySummaries,
  subjectNames,
  writing = false,
  eventParams = {
    action: "msl_link_click",
    category: "msl_ichiran",
    label: "msl_mansion_title",
    value: 1,
  },
}: MansionCardProps) => {
  const clickMansion = () => {
    event(eventParams);
  };
  const linkProps = {
    href: "/mansions/[uniqueCode]",
    as: buildLibraryPagePath("mansions", uniqueCode),
  };
  const layout = display.layoutName;
  const address = compact([
    prefecture.name,
    city.name,
    town.name,
    street?.name,
    blockNumber,
  ]).join("");
  const priceRange = uniq([
    yenFormat(display.minPrice),
    yenFormat(display.maxPrice),
  ]).join(" ~ ");
  const exclusiveArea = uniq([
    display.minExclusiveArea + "㎡",
    display.maxExclusiveArea + "㎡",
  ]).join(" ~ ");
  const hasSafty = Date.parse(constructedIn) < Date.parse("1983-06-01");
  const constructedDate = `${format(
    Date.parse(constructedIn),
    "yyyy年MM月"
  )} (築${age}年)`;

  const accessText = access.map(
    ({ stationName, stationLineName, stationWalkingMinutes }, i) => {
      return (
        <span
          key={i}
          className={
            subjectNames &&
            (subjectNames.includes(stationName) ||
              subjectNames.includes(stationLineName))
              ? styles.highlight
              : ""
          }
        >
          {`${stationLineName}「${stationName}駅」`}
          {stationWalkingMinutes && `徒歩${stationWalkingMinutes}分`}
        </span>
      );
    }
  );

  const mansionScale = (() => {
    if (!unitAmount) return "";
    else if (unitAmount > 100) return "大規模";
    else if (unitAmount > 20) return "中規模";
    else return "小規模";
  })();

  const accessConvenience = (() => {
    const closestStationWalkingMinutes = Math.min(
      ...access.map(({ stationWalkingMinutes }) =>
        Number(stationWalkingMinutes)
      )
    );

    if (closestStationWalkingMinutes <= 5)
      return "ロケーションにも優れています。";
    else if (closestStationWalkingMinutes <= 10)
      return "主要都市へのアクセスに便利です。";
    else return "";
  })();

  // 間取りの小さい順に並び替えた MansionSalesHistorySummary
  const sortedSummaries = sortBy(salesHistorySummaries, [
    "numberOfRooms",
    ({ layoutKind }) => layoutKind.sort,
  ]);

  // 小さい順に並び替えた全 MansionSalesHistorySummary の専有面積
  const sortedAreas = salesHistorySummaries
    .reduce(
      (areas, { exclusiveAreas }) => areas.concat(exclusiveAreas),
      [] as number[]
    )
    .sort((a, b) => a - b);

  const minLayoutSummary = sortedSummaries[0];
  const maxLayoutSummary = sortedSummaries[sortedSummaries.length - 1];
  const minExclusiveArea = sortedAreas[0];
  const maxExclusiveArea = sortedAreas[sortedAreas.length - 1];

  const layoutRangeText = uniq([
    minLayoutSummary.layout,
    maxLayoutSummary.layout,
  ]).join("～");

  function createSuitableUseText(
    numberOfRooms: number,
    layoutKindSlug: string
  ): string {
    switch (numberOfRooms) {
      case 1:
        if (layoutKindSlug === "r" || layoutKindSlug === "k") {
          return "一人暮らしを検討している方";
        }
        return "部屋を活用する一人暮らし・二人暮らしを検討している方";
      case 2:
        return "ゆったりとした二人暮らしを検討している方や、お子様がお一人いるご家庭";
      case 3:
        return "お子様が二人いるご家庭で広々と部屋をご活用したい場合";
      default:
        return "お子様が二人以上いて子供部屋が欲しいとお考えの場合や住むご予定の人数が多い場合";
    }
  }

  const suitableUseRangeText = uniq([
    createSuitableUseText(
      minLayoutSummary.numberOfRooms,
      minLayoutSummary.layoutKind.slug
    ),
    createSuitableUseText(
      maxLayoutSummary.numberOfRooms,
      maxLayoutSummary.layoutKind.slug
    ),
  ]).join("～");

  const exclusiveAreaRangeText = uniq([
    `${minExclusiveArea}㎡`,
    `${maxExclusiveArea}㎡`,
  ]).join("～");

  return (
    <div className={styles.container}>
      <div className={styles.pcOnly}>
        <div className={styles.block}>
          <Link {...linkProps} prefetch={false} passHref>
            <button className={styles.name} onClick={() => clickMansion()}>
              <h3>{name}</h3>
            </button>
          </Link>
          <div className={styles.columns}>
            <ul className={styles.table}>
              <li className={styles.item}>
                <p className={styles.label}>物件名</p>
                <p className={styles.content}>{name}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>中古販売価格</p>
                <p className={styles.content}>{priceRange}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>所在地</p>
                <p className={styles.content}>{address}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>沿線・駅</p>
                <p className={styles.content}>{accessText}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>占有面積</p>
                <p className={styles.content}>{exclusiveArea}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>間取り</p>
                <p className={styles.content}>{layout}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>築年月</p>
                <p className={styles.content}>{constructedDate}</p>
              </li>
              <li className={styles.item}>
                <p className={styles.label}>総戸数</p>
                <p className={styles.content}>
                  {unitAmount ? `${unitAmount}戸` : "-"}
                </p>
              </li>
            </ul>
          </div>
          {pageView && (
            <p className={styles.page_view}>
              ※ {pageView}人がこの物件を閲覧しました。
            </p>
          )}
          {writing && (
            <div className={classnames(tableStyles.tableWrapper, "m_b_30px")}>
              <table className={tableStyles.table}>
                <tbody>
                  <tr>
                    <td>
                      <p>
                        {name}は{address}に位置する総戸数{unitAmount}戸の
                        {mansionScale}マンションです。
                      </p>
                      <p>
                        最寄駅は{accessText}で、{accessConvenience}
                        {age <= 16 ? (
                          <>
                            {name}の築年月は{constructedDate}で、
                            {age > 5 ? "比較的" : ""}新しい物件となっています。
                            {name}の販売価格は{priceRange}です。
                          </>
                        ) : (
                          <>
                            {name}の築年月は{constructedDate}であり、 販売価格は
                            {priceRange}です。
                          </>
                        )}
                      </p>
                      <p>
                        {name}間取りには{layoutRangeText}があり、
                        {suitableUseRangeText}
                        におすすめのマンションです。
                      </p>
                      <p>
                        広さでみると{exclusiveAreaRangeText}
                        の物件があり、{name}
                        では表内の色がついている広さの部屋が存在しています。
                      </p>
                      {hasSafty && (
                        <p>
                          {name}
                          の耐震性は新耐性基準を採用しており、中規模の地震（震度5強程度）でほとんど損傷せず、大規模な地震震度6強～7程度の揺れでも倒壊しないような構造基準が適用されており、安心して住める基準を満たしています。
                        </p>
                      )}
                      <p>
                        <_Link
                          text="Housii"
                          linkProps={{ href: "https://ieul.jp/buy/" }}
                          eventParams={eventParams}
                        />
                        に登録することで、{name}
                        の最新情報をお届けいたします。
                      </p>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          )}
          <Link {...linkProps} prefetch={false} passHref>
            <button onClick={() => clickMansion()} className={styles.link}>
              この物件の詳細を見る
            </button>
          </Link>
        </div>
      </div>
      <div className={styles.spOnly}>
        <Link {...linkProps} prefetch={false} passHref>
          <button onClick={() => clickMansion()} className={styles.card}>
            <h3>{name}</h3>
            <ul>
              <li>{priceRange}</li>
              <li>{exclusiveArea}</li>
              <li>{layout}</li>
              <li>{constructedDate}</li>
              <li>{unitAmount}</li>
            </ul>
            <p>
              {address}
              <br />
              {accessText}
            </p>
          </button>
        </Link>
      </div>
    </div>
  );
};

export default MansionCard;
