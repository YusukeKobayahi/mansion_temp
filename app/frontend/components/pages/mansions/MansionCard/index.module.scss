@import "~/styles/variables";
@import "~/styles/utils";

.container {
  position: relative;
  width: 100%;
  height: 100%;
  border: 1px solid #e5e5e5;
  border-radius: 4px;
  background-color: #fff;
}

.link {
  width: 100%;
  height: 100%;
  overflow: hidden;
  padding: 20px;
  display: flex;
  flex-direction: column;
  background-color: unset;
  text-align: left;

  @include screenMd {
    padding: 15px;
  }
}

.name {
  overflow: hidden;
  width: 100%;
  display: inline-block;
  font-size: 18px;
  color: $color-main;
  text-decoration: underline;
  line-height: 1.4;
  min-height: 50px;

  @include screenMd {
    min-height: auto;
    font-size: 14px;
  }
}

.date {
  display: inline-block;
  color: $color-text-main;
  font-size: 14px;
  line-height: 1.5;

  @include screenMd {
    font-size: 12px;
  }
}

.salesInfoArea {
  margin: 10px 0;
  line-height: 1.5;
  color: #373a3c;
  font-size: 12px;

  > p {
    line-height: 2;
  }

  ul {
    list-style-type: disc;
    margin-left: 25px;

    li {
      color: #373a3c;
    }
  }
}

.salesInfo {
  font-size: 14px;
}

.subInfo {
  margin-top: auto;
  line-height: 1.5;
  font-size: 14px;
  color: #9d9d9d;
}
