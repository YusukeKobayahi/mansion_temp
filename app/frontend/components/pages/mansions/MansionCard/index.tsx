import Link from "next/link";
import { event } from "~/lib/gtag";
import { LayoutSummary } from "~/lib/mansions/SalesSummaryBuilder";
import compact from "lodash/compact";
import styles from "./index.module.scss";
import { GoogleAnalyticsEvent } from "~/lib/types";

export interface MansionCardProps {
  href?: string;
  name?: string;
  address?: string;
  salesInfo?: LayoutSummary[];
  constructedDate?: string;
  access?: string;
  eventParams?: GoogleAnalyticsEvent;
}

const MansionCard: React.FC<MansionCardProps> = ({
  href,
  name,
  constructedDate,
  salesInfo,
  address,
  access,
  eventParams,
}: MansionCardProps) => {
  const clickLink = () => {
    if (eventParams) event(eventParams);
  };
  return (
    <div className={styles.container}>
      <Link href={"/mansions/[uniqueCode]"} as={href} prefetch={false} passHref>
        <button className={styles.link} onClick={clickLink}>
          <h3 className={styles.name}>
            {name}
            <br />
            <span className={styles.date}>{constructedDate}</span>
          </h3>
          <div className={styles.salesInfoArea}>
            <p>販売価格</p>
            <ul>
              {salesInfo
                ? salesInfo.map((info, i) => (
                    <li key={i} className={styles.salesInfo}>
                      {compact([info.layout, info.info]).join(" ")}
                    </li>
                  ))
                : null}
            </ul>
          </div>
          <p className={styles.subInfo}>
            <span>{address}</span>
            <br />
            {access ? <span>{access}</span> : null}
          </p>
        </button>
      </Link>
    </div>
  );
};

export default MansionCard;
