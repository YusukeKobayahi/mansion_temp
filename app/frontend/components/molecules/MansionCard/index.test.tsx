import { render } from "@testing-library/react";
import MansionCard from ".";

describe("components/commons/MansionCard", () => {
  it("正常に表示される", () => {
    const mansion = {
      uniqueCode: "868b97434821477f8dbe33b5",
      name: "レーベンリヴァーレエイセスモバード",
      age: 10,
      constructedIn: "2012-03-01",
      unitAmount: 86,
      prefecture: {
        name: "神奈川県",
      },
      city: {
        name: "鎌倉市",
      },
      town: {
        name: "台",
      },
      street: {
        name: "５丁目",
      },
      blockNumber: "３－４",
      display: {
        minPrice: 30800000,
        maxPrice: 39800000,
        minExclusiveArea: 66.72,
        maxExclusiveArea: 80.47,
        layoutName: "2SLDK、3LDK",
      },
      access: [
        {
          id: "18",
          stationId: 8030,
          stationLineName: "東海道線",
          stationName: "大船",
          stationWalkingMinutes: 15,
          lineId: 427,
        },
      ],
      salesHistorySummaries: [
        {
          layout: "1LDK",
          numberOfRooms: 1,
          layoutKind: {
            slug: "ldk",
            sort: 7,
          },
          exclusiveAreas: [55.04, 59.42],
        },
      ],
    };
    const subjectNames = ["東海道線"];
    const { asFragment } = render(
      <MansionCard {...mansion} subjectNames={subjectNames} writing={false} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
