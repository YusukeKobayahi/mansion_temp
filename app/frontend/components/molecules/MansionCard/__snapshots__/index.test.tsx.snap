// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/commons/MansionCard 正常に表示される 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="pcOnly"
    >
      <div
        class="block"
      >
        <button
          class="name"
          href="/mansions/868b97434821477f8dbe33b5"
        >
          <h3>
            レーベンリヴァーレエイセスモバード
          </h3>
        </button>
        <div
          class="columns"
        >
          <ul
            class="table"
          >
            <li
              class="item"
            >
              <p
                class="label"
              >
                物件名
              </p>
              <p
                class="content"
              >
                レーベンリヴァーレエイセスモバード
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                中古販売価格
              </p>
              <p
                class="content"
              >
                3,080万円 ~ 3,980万円
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                所在地
              </p>
              <p
                class="content"
              >
                神奈川県鎌倉市台５丁目３－４
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                沿線・駅
              </p>
              <p
                class="content"
              >
                <span
                  class="highlight"
                >
                  東海道線「大船駅」徒歩15分
                </span>
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                占有面積
              </p>
              <p
                class="content"
              >
                66.72㎡ ~ 80.47㎡
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                間取り
              </p>
              <p
                class="content"
              >
                2SLDK、3LDK
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                築年月
              </p>
              <p
                class="content"
              >
                2012年03月 (築10年)
              </p>
            </li>
            <li
              class="item"
            >
              <p
                class="label"
              >
                総戸数
              </p>
              <p
                class="content"
              >
                86戸
              </p>
            </li>
          </ul>
        </div>
        <button
          class="link"
          href="/mansions/868b97434821477f8dbe33b5"
        >
          この物件の詳細を見る
        </button>
      </div>
    </div>
    <div
      class="spOnly"
    >
      <button
        class="card"
        href="/mansions/868b97434821477f8dbe33b5"
      >
        <h3>
          レーベンリヴァーレエイセスモバード
        </h3>
        <ul>
          <li>
            3,080万円 ~ 3,980万円
          </li>
          <li>
            66.72㎡ ~ 80.47㎡
          </li>
          <li>
            2SLDK、3LDK
          </li>
          <li>
            2012年03月 (築10年)
          </li>
          <li>
            86
          </li>
        </ul>
        <p>
          神奈川県鎌倉市台５丁目３－４
          <br />
          <span
            class="highlight"
          >
            東海道線「大船駅」徒歩15分
          </span>
        </p>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
